working/333 - shared memory multithreading/proposal.md

```diff
@@ -842,9 +594,7 @@ class Isolate {
   /// Shared isolate contains a copy of the
   /// global `shared` state of the current isolate and does not have any
   /// non-`shared` state of its own. An attempt to access non-`shared` static variable throws [IsolationError].
-  ///
-  /// If [task] is not [Shareable] then [ArgumentError] is thrown.
-  static Future<S> runShared<S extends Shared>(S Function() task);
+  static Future<S> runShared<S>(S Function() task);
```

Comment:
Could you explain a bit more about how the ffi interop would actual work in this model? In particular, do callbacks running as "shared" capture their closure context?

I'm thinking about use cases like this:

You want to implement an Objective-C delegate in Dart to limit the number of HTTP redirects in your HTTP library. The delegate method looks like this and can be invoked on any thread:

```objc
- (void)URLSession:(NSURLSession *)session 
              task:(NSURLSessionTask *)task 
willPerformHTTPRedirection:(NSHTTPURLResponse *)response 
        newRequest:(NSURLRequest *)request 
 completionHandler:(void (^)(NSURLRequest *))completionHandler;
```

Your Dart code could look something like:

```dart
Future<String> doGet(Uri uri, int maxRedirects) {
  ...
  var numRedirects = 0;
  delegate.URLSession_task_willPreformHTTPRedirection_... = 
        (session, task, response, request, completionHandler) {
    numRedirects += 1; // Is this accessable?
    if (numRedirects <= maxRedirects) {
      ...
    } else {
      ...
    }
  }
}
```

Would `numRedirects` be available in the callback? If not, how would you make that sort of non-global state available?

Reply:
It looks like @aam 's prototype of `isolateGroupShared` does do closure capture so that would solve this problem.

Reply:
I think this code is okay because integers are trivially shareable. 

See my answer to @mkustermann above: https://github.com/dart-lang/language/pull/4095#discussion_r1778662388

However if the state is more complex and not trivially shareable then callback would need to be explicit and enter some isolate either by sending a task to some isolate or by using `Isolate.runSync` to enter isolate synchronously.

Reply:
> will the mutation of numRedirects in the callback we seen by other threads

There is one instance of `numRedirects` per invocation of `doGet`. So isolateGroupShared handler created in that `doGet` will get access to only that one `numRedirects` (via captured context).

Reply:
> In my example, is `numRedirects` shared between threads i.e. will the mutation of `numRedirects` in the callback we seen by other threads?

Yes, it will be shared. If multiple threads invoke the same callback they will attempt to mutate the same variable. 

(Which also means that if you expect that you should use an atomic variable).